[fusion] Migrate away from CustomFuseGraph

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#72 [fusion] Migrate away from CustomFuseGraph**

bwasti committed:
[fusion] Migrate away from CustomFuseGraph

bwasti committed:
Update on "[fusion] Migrate away from CustomFuseGraph"

[fusion] Migrate away from CustomFuseGraph

gh-metadata: pytorch tvm 72 gh/bwasti/46/head

bwasti commented:
this depends on https://github.com/pytorch/pytorch/pull/23210

bwasti committed:
Update on "[fusion] Migrate away from CustomFuseGraph"

[fusion] Migrate away from CustomFuseGraph

gh-metadata: pytorch tvm 72 gh/bwasti/46/head

wanchaol reviewed (commented):
I tried to use this to compare the tvm fusion group we generated before and after, and looks like we are pulling in redundant constant nodes again and again, which I think we shouldn't. Taking a example:

```
with tvm::CompilationGroup_1 = graph(%0 : Tensor,
      %1 : Float(*, *, *),
      %2 : Float(*)):
  %11 : int[] = prim::Constant[value=[0]]()
  %3 : int[] = prim::Constant[value=[1]]()
  %4 : int[] = prim::Constant[value=[0]]()
  %5 : int[] = prim::Constant[value=[1]]()
  %6 : bool = prim::Constant[value=0]()
  %7 : int[] = prim::Constant[value=[0]]()
  %8 : int = prim::Constant[value=1]()
  %9 : bool = prim::Constant[value=1]()
  %x1.1 : Tensor = aten::_convolution(%0, %1, %2, %3, %4, %5, %6, %7, %8, %6, %6, %9) # code/my_noqqq.py:377:8
  return (%x1.1)
with tvm::CompilationGroup_2 = graph(%0 : Tensor,
      %1 : Float(*, *, *),
      %2 : Float(*)):
  %11 : int[] = prim::Constant[value=[0]]()
  %3 : int[] = prim::Constant[value=[1]]()
  %4 : int[] = prim::Constant[value=[0]]()
  %5 : int[] = prim::Constant[value=[1]]()
  %6 : bool = prim::Constant[value=0]()
  %7 : int[] = prim::Constant[value=[0]]()
  %8 : int = prim::Constant[value=1]()
  %9 : bool = prim::Constant[value=1]()
  %x2.1 : Tensor = aten::_convolution(%0, %1, %2, %3, %4, %5, %6, %7, %8, %6, %6, %9) # code/my_noqqq.py:380:8
  return (%x2.1)
```
I thought that the graph executor run constant pooling somewhere but it turned out not running it properly.. this will probably need to do more passes after we generate the fusion group like the graphFuser did:

```
  // After FuseGraph some common subexpressions may come back
  EliminateCommonSubexpression(graph);
  // We might have emitted a fair amount of useless shape propagating code, so
  // remove it
  EliminateDeadCode(graph);
  // Improve the quality of shape propagation code that was left
  PeepholeOptimizeShapeExpressions(graph->block());
```

Also, we might think of just adding constants as arguments rather than pulling it into the fusion group, this way it's easier to run constant pooling stuff I think.

t-vi commented:
So I'm ignorant of the underlying goal of not using the graph fuser, but the constants seem to be an indication that it might be nice to try to use as much of PyTorch's capabilities as possible.

In contrast to the current master torch_tvm (see #55) and apparently also this PR, the PyTorch fuser treats constants more delicately: It will _not_ fuse them on their own but instead copies constant inputs of operation nodes to be fused into the fusion group and reconnects the fused operation to these copies.

bwasti commented:
@wanchaol @t-vi good points on being careful with `prim::Constant`, I will update that to be consistent with the current graph fusion code.

@t-vi The issue is that GraphFuser doesn't (and probably shouldn't) handle the more complex cases of aliasing and control flow.  For the sake of Relay lowering, we will want to have loops and conditionals as well as tensor views well supported to generate the best code.  I put up this diff as the first step in many to get that all working 😃

t-vi commented:
Good point about the control flow which would not be needed for basic graph fusion. I'm sure that people will come up with other uses for control-flow capable fusers if they get them... :)

bwasti committed:
Update on "[fusion] Migrate away from CustomFuseGraph"

[fusion] Migrate away from CustomFuseGraph

gh-metadata: pytorch tvm 72 gh/bwasti/46/head

kimishpatel commented on torch_tvm/fusion_pass.cpp:
```diff
@@ -0,0 +1,107 @@
+    if (auto group = tryMerge(consumer, input->node(), aliasDb)) {
+      // we successfully merged, so the new group's `inputs` may have
+      // changed. So rescan the new group for more merging opportunities.
+      return group.value()->reverseIterator();
```
has_value check?

t-vi replied:
if it has not, the if will evaluate to false.

bwasti committed:
Update on "[fusion] Migrate away from CustomFuseGraph"

[fusion] Migrate away from CustomFuseGraph

gh-metadata: pytorch tvm 72 gh/bwasti/46/head

yinghai commented:
@bwasti Could you add some reasoning on this PR? Like why do we need to migrate away from custom graph fuse.

bwasti committed:
Update on "[fusion] Migrate away from CustomFuseGraph"

[fusion] Migrate away from CustomFuseGraph

gh-metadata: pytorch tvm 72 gh/bwasti/46/head

bwasti committed:
Update on "[fusion] Migrate away from CustomFuseGraph"

[fusion] Migrate away from CustomFuseGraph

gh-metadata: pytorch tvm 72 gh/bwasti/46/head

zrphercule commented:
Is this pr ready to be merged now?

bwasti commented:
@yinghai I added some notes above in response to @t-vi who had a similar question.  Largely it comes down to proper handling of alias information and enabling control flow in the future

@zrphercule I believe so, yes.  I will add a few more tests on coverage

yinghai commented on torch_tvm/fusion_pass.h:
```diff
@@ -0,0 +1,7 @@
+#include <torch/csrc/jit/passes/graph_fuser.h>
+#include <torch/csrc/jit/passes/utils/subgraph_utils.h>
+
+void FuseSupportedOps(std::shared_ptr<torch::jit::Graph> graph);
```
Actually, is this function TVM specific? If not, can we move it to pytorch repro?

bwasti replied:
that's a good catch.  For the sake of efficiency we can "stage" a lot of simple functionality like this here and upstream it to PyTorch at some point.

I'd like to prioritize flushing out functionality before trying to standardize it into an API.  That being said, I've tried to make it very copy-and-paste-able 😛

jackm321 replied:
>very copy-and-paste-able

Lol sounds useful

yinghai replied:
> That being said, I've tried to make it very copy-and-paste-able

That's something that can be landed directly into Pytorch in the spirit of hackability of PyTorch. Lol

t-vi commented on torch_tvm/fusion_pass.cpp:
```diff
@@ -0,0 +1,134 @@
+bool canHandle(Block* block, AliasDb& aliasDb) {
+  for (Node* node : block->nodes()) {
+    if (!canHandle(node, aliasDb)) {
+      return false;
```
The traditional fuser works on blocks, traversing them in the fusion pass. This here seems more about whether we can fuse everything in the block in order to fuse the entire block-using control flow statement.

jackm321 replied:
If I understand what you mean, that's what I'm saying. It looks the the traditional fuser recurses on sub-blocks fusing what it can within them whereas this will only fuse a sub-block if all nodes within it are fusable (recursively). So I think the traditional fuser could for example fuse most of a loop body while leaving the any unsupported nodes in the loop body unfused wheras this I think will try to fuse the entire loop or nothing.

bwasti replied:
yep, this means any single fusion attempt will be all or nothing.  We can add attempts to recursively fuse on blocks that weren't fused on a previous attempt (in a later diff maybe?) to recover the CustomGraphFuse behavior
